### ISSUE TITLE
How caffe calculate total loss when training faster rcnn by Approximate joint training way?

### ISSUE THREAD

**Issue by XiaotaoChen** [rbgirshick/py-faster-rcnn#643]
Hi,
   recently, i read faster rcnn's source code. i have two questions as follows,

1. **how caffe calculate total loss among rpn_cls_loss, rpn_loss_bbox,loss_cls, loss_bbox?**

   i haven't found the 'sum loss' code in faster rcnn's source code. Can caffe auto find the multiple loss in the defined net  and sum them weight by each loss_weight? How is it work? Can anyone explain it and tell me this part code in caffe cource?

2. **how the network do the backwards according to the total_loss?**

   after got the total loss. there are four branches(layers which top-out is rpn_cls_loss, rpn_loss_bbox, loss_cls, loss_bbox) needed to do backwards ?  Firstly, how to allocate the total loss to four branches ? is each branch' loss= total-loss/4 and then to do backwards independently ? Secondly, the four branches are above on VGG16. So if The four branches do backwards independently, The paramaters in VGG16 will be updated four times . Is my guess correct ?

3. **Question about the network defined by the faster_rcnn_end2end/train.prototxt. why haven't set propagate_down in the two layers for bbox regression which type is SmoothL1Loss ? However, the two layers for cls-loss both had set propagate_down? **

   the bottom data of 'rpn_labels' in 'rpn_loss_cls' layer and the  bottom data of 'rpn_bbox_targets' in 'rpn_loss_bbox' layer are both come from 'rpn-data' layer, and the 'rpn_loss_cls' and 'rpn_bbox_targets' layers both are calculate for loss, What is the different settting's reason ? According to the explaination of approximae joint training, it say just use the cls-loss to do backwards, it's loss hasn't include bbox-loss, So the net hasn't setting the propagate_down = 1 to allow the layer's loss backward ? What is the default setting about 'propagate_down' ?

**Comment by makefile** [comment]
1. track from `solver.cpp` we can see that loss is calculated by forward funtion in `net.cpp` which sum up the loss of each layer from bottom to top layer. Only Loss layer has meaningful loss value. The four type `rpn_cls_loss, rpn_loss_bbox,loss_cls, loss_bbox` is calculated and weighted by each layer.Since the loss value has already been weighted,so caffe just sum them up.
2. The 'total_loss' is not a layer,so there is no handling about 'total_loss' . Branches are backpropgated in divide-and-sum way. The paramaters will be updated only once at every time of backward.
3. The `propagate_down` param indicates whether to backward in that layer,and the default value is true.Since `rpn_labels` and `labels` bottom blob is fixed, there is no need to learn, then set `propagate_down: false`.
